### PR TITLE
Make KubeSim.Run() and AddSubmitter() thread-safe

### LINKED ISSUE #151: Make KubeSim.AddSubmitter() and Run thread-safe

### PR DESCRIPTION
**Pull request by ordovicia**
`KubeSim.AddSubmitter()` can be called from the outside in parallel with `Run()`, so these two methods should be thread-safe.
With this PR, for the storage of submitters, a `KubeSim` instance uses `sync.Map` instead of raw `map[string]Submitter`.
The `sync.Map` is wrapped in `SubmitterMap` struct for the type-safety.

Fixes #151

### CODE REVIEW

File: pkg/kubesim/submitter_map/submitter_map.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+}
+
+// Delete deletes and returns the submitter associated with the key, or returns nil if no value is
+// present.
+// The second return value indicates whether the submitter was found.
+func (s *SubmitterMap) Delete(key string) (subm.Submitter, bool) {
+	val, ok := s.Load(key)
+	if ok {
+		s.inner.Delete(key)
+		s.size--
```
Comment (dtaniwaki) [inline pkg/kubesim/submitter_map/submitter_map.go]: Is `size` thread-safe?

**Comment by ordovicia** [comment]
> Is size thread-safe? I'm not that familiar with Go runtime, but I think there's a possibility to access size in parallel unless a lock object is got.

I realized `SubmitterMap` is entirely not thread-safe. I will fix `SubmitterMap` if making it thread-safe is necessary.

> By the way, do we really need to make submitters thread-safe? Is adding submitters in the running phase expected usage? If yes, the change of this PR is necessary, but if not, I think we can just make calling AddSubmitter failed after the simulator started to run.

I wrote this PR simply because the current implementation is in race condition.

I'm not sure the current design for adding submitters is the best. Actually, we can completely remove `AddSubmitter()`. I.e. If submitters are statically registered to `KubeSim`, we can modify `NewKubeSim*()` functions to take the submitters. Multiple submitters are passed as a list, or are gathered into another `Submitter` by using the composite pattern. Even if submitters are dynamically registered, this registration can be handled in a "composite" submitter.
To me, modifying `NewKubeSim*()` to statically take a single submitter looks the best. By this change, we can convey the intent, by the syntactic interface, that `KubeSim` doesn't allow dynamic submitter registration. Also we can separate the interest on multiple submitters and the dynamic registration from `KubeSim`. We can provide a common composite submitter in this package.

@dtaniwaki How do you think?

**Comment by dtaniwaki** [comment]
I think your idea of `passing multiple submitters as an argument` is nice. I think adding submitters dynamically is a very rare case and it can be covered by changing the behavior of a submitter dynamically by themselves.

I'm not sure if forcing users to set only one submitter and using composite pattern is intuitive and it's too much if only a few submitters are necessary.
However, we should provide a common composite submitter as you said above, let's make one in the near future.

**Comment by ordovicia** [comment]
Closed in favor of #205

**Pull request closed without merging**